api-select.go

```diff
@@ -0,0 +1,637 @@
+		contentMD5Base64: sumMD5Base64(selectReqBytes),
+		contentSHA256Hex: sum256Hex(selectReqBytes),
+		contentBody:      bytes.NewReader(selectReqBytes),
+		contentLength:    int64(len(selectReqBytes)),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp, bucketName, "")
+	}
+	// Process needs to return a eventStream to the client
+	// Wrap this in a Go routine so as to allow the streaming
+	streamer := &Events{
+		Response: make(chan interface{}, 2),
+		lock:     make(chan bool, 2),
+	}
+	go func() {
+		processBinary(resp.Body, streamer)
+	}()
+	return streamer, nil
+}
+
+// processBinary is the main function that decodes the large byte array into
+// several events that are sent through the eventstream.
+func processBinary(message io.ReadCloser, streamer *Events) {
+	defer message.Close()
+	var partialRecordVal []byte
+	lenOfTrim := 0
+	for {
+		// select block for catching a user request for closing.
+		select {
+		// Logic is basically that the .Close() method will close the streamer.lock
+		// channel, Therefore if Ok is false(the channel is closed) then we can
+		// close our response channel and return. The .Close() method blocks on the
+		// streamer.Response channel being closed, so that it is guaranteed that this
+		// method has exited since it is a defered close.
+		case _, ok := <-streamer.lock:
+			if !ok {
+				defer close(streamer.Response)
+				return
+			}
+		default:
+			var prelude preludeInfo
+			var headers *headerInfo
+			// Create CRC code
+			crc := crc32.New(crc32.IEEETable)
+			crcReader := io.TeeReader(message, crc)
+			var err error
+			var payload []byte
+			// Extract the prelude(12 bytes) into a struct to extract relevant
+			// information.
+			prelude, err = processPrelude(crcReader, crc)
+			if err != nil {
+				if streamer.err == nil {
+					streamer.err = err
+				}
+				close(streamer.Response)
+				return
+			}
+			// Extract the headers(variable bytes) into a struct to extract relevant
+			// information
+			if prelude.headerLen > 0 {
+				headers, err = extractHeader(io.LimitReader(crcReader, int64(prelude.headerLen)))
+				if err != nil {
+					if streamer.err == nil {
+						streamer.err = err
+					}
+					close(streamer.Response)
+					return
+				}
+			}
+			// Calculate the length of the payload so that the appropriate amount of
+			// bytes can be read
+			payloadLen := calcPayloadLen(prelude)
+			// Checks that there is a payload and if there is, then extracts it.
+			if payloadLen > 0 {
+				payload, err = extractPayload(io.LimitReader(crcReader, int64(payloadLen)), payloadLen)
+				if err != nil {
+					if streamer.err == nil {
+						streamer.err = err
+					}
+					close(streamer.Response)
+					return
+				}
+				// Special case for dealing with messages that are partial.
+				if isPartial(payload) {
+					payload, partialRecordVal, lenOfTrim = processPartial(payload)
+					processEvent(payload, headers, streamer)
```

Comment:
This code looks a bit repetitive - it looks like it should be possible to refactor this so that there is only a single call to 
extractPayload, processPartial, processEvent within a loop. That will be easier to read and maintain.